Store Wi-Fi firmware patch in external flash on nRF7002 DK

ArekBalysNordic opened the pull request:
In this PR there are the following changes:

- Enabled storing the Wi-Fi firmware patch in external flash. Removed notes, and todo sections.
- Updated the `nrf70-fw-patch-ext-flash` snippet to work with nRF7002DK (Enabled SPI).
- Added build configuration for Wi-Fi Shell sample to store the patch in the external flash.
- Switched to storing the patch in an external flash in all Matter samples by default.
- Updated README files to inform about the change.
- Added release notes and migration guide entries.

This is a placeholder for testing and it is going to be merged once nRFUtil supports flashing the external flash through SPI.

@krish2718, @sachinthegreen, @rado17 and @nrfconnect/ncs-wifi-doc could you please review the changes related to the Wi-Fi?

NordicBuilder commented:
## CI Information

<sup>_To view the history of this post, clich the 'edited' button above_</sup>
Build number: 1
### Inputs:
#### Sources:
<details><summary> more details</summary>
</details>

#### Github labels
| Enabled | Name | Description |
| -- | -- | -- |
|| ci-disabled | _Disable the ci execution_ | 
|| ci-all-test | _Run all of ci, no test spec filtering will be done_ | 
|| ci-force-downstream | _Force execution of downstream even if twister fails_ | 
|| ci-run-twister | _Force run twister_ | 
|| ci-run-zephyr-twister | _Force run zephyr twister_ | 

<details><summary> List of changed files detected by CI (0)</summary>

```
```
</details>

### Outputs:

#### Toolchain
Version: 
Build docker image: 

#### Test Spec & Results: :white_check_mark: Success; :x: Failure; :orange_circle: Queued; :yellow_circle: Progress; :white_medium_square: Skipped; :warning: Quarantine
- :x: Toolchain
- :x: Build twister
- :x: Integration tests



*Note: This message is automatically posted and updated by the CI*

krish2718 commented:
> After documentation is built, you will find the preview for this PR [here](https://ncsdoc.z6.web.core.windows.net/PR-20528).
> 
> Preview links for modified nRF Connect SDK documents:
> 
> https://ncsdoc.z6.web.core.windows.net/PR-20528/nrf/app_dev/device_guides/nrf70/nrf70_fw_patch_update.html https://ncsdoc.z6.web.core.windows.net/PR-20528/nrf/releases_and_maturity/migration/migration_guide_3.0.html https://ncsdoc.z6.web.core.windows.net/PR-20528/nrf/releases_and_maturity/releases/release-notes-changelog.html https://ncsdoc.z6.web.core.windows.net/PR-20528/nrf/applications/matter_bridge/doc/matter_bridge_description.html https://ncsdoc.z6.web.core.windows.net/PR-20528/nrf/samples/matter/light_bulb/README.html https://ncsdoc.z6.web.core.windows.net/PR-20528/nrf/samples/matter/light_switch/README.html https://ncsdoc.z6.web.core.windows.net/PR-20528/nrf/samples/matter/lock/README.html https://ncsdoc.z6.web.core.windows.net/PR-20528/nrf/samples/matter/template/README.html https://ncsdoc.z6.web.core.windows.net/PR-20528/nrf/samples/matter/thermostat/README.html https://ncsdoc.z6.web.core.windows.net/PR-20528/nrf/samples/matter/window_covering/README.html

The links are not working, are they expired? Do we need to re-run the doc action?

ArekBalysNordic commented:
> The links are not working, are they expired? Do we need to re-run the doc action?

@krish2718 Should work now, please check. :)

peknis commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -281,6 +281,8 @@ Matter Bridge
 * Updated by enabling Link Time Optimization (LTO) by default for the ``release`` configuration.
 * Removed support for the nRF54H20 devices.
+* Enabled Link Time Optimization (LTO) by default for the ``release`` configuration.
+* Enabled the default storage of the Wi-Fi firmware patch in external flash.
```
Suggesting to reorganize the items like this:

* Updated:

  * By enabling Link Time Optimization...
  * By enabling the default storage...

* Removed support for the nRF54H20 devices.

peknis commented on doc/nrf/releases_and_maturity/migration/migration_guide_3.0.rst:
```diff
@@ -49,6 +49,59 @@ nRF5340 Audio applications
+
+      * The Wi-Fi firmware patch is now stored in the external flash memory by default for the nRF7002 DK and nRF5340 DK with the nRF7002 EK shield attached.
+        It means that the default partitions for the Wi-Fi configuration have been changed for these boards.
+        The Wi-Fi firmware patch is located in the ``nrf70_wifi_fw`` partition within the external flash memory space, and the ``mcuboot_primary_2`` partition is used for the updating the patch pruposes.
```
```suggestion
        The Wi-Fi firmware patch is located in the ``nrf70_wifi_fw`` partition within the external flash memory space and the ``mcuboot_primary_2`` partition is used for the updating the patch purposes.
```

peknis commented on doc/nrf/releases_and_maturity/migration/migration_guide_3.0.rst:
```diff
@@ -49,6 +49,59 @@ nRF5340 Audio applications
+        It means that the default partitions for the Wi-Fi configuration have been changed for these boards.
+        The Wi-Fi firmware patch is located in the ``nrf70_wifi_fw`` partition within the external flash memory space, and the ``mcuboot_primary_2`` partition is used for the updating the patch pruposes.
+
+        To keep storing the Wi-Fi firmware patch in the internal flash memory, do the following actions:
```
```suggestion
        To keep storing the Wi-Fi firmware patch in the internal flash memory, perform the following actions:
```

peknis commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -650,6 +652,15 @@ Matter samples
   * Removed support for nRF54H20 devices.
   * Updated the API of ``AppTask``, ``BoltLockManager``, and ``AccessManager`` to provide additional information for the ``LockOperation`` event.
 
+* Updated all Matter samples that runs Wi-Fi stack to store the Wi-Fi firmware patch in external flash.
```
Suggesting that you place this entry under the first Added bullet.

ArekBalysNordic replied:
ok, moved to the `Updated` bullet.

peknis commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -642,6 +646,14 @@ Matter samples
 
     * The documentation with instructions on how to build the sample on the nRF54L15 DK with support for Matter OTA DFU and DFU over Bluetooth SMP, and using internal RRAM only.
     * Link Time Optimization (LTO) to be enabled by default for the ``release`` configuration and ``nrf7002dk/nrf5340/cpuapp`` build target.
+    * All Matter samples that runs Wi-Fi stack to store the Wi-Fi firmware patch in external flash.
```
```suggestion
    * Updated all Matter samples that run Wi-Fi stack to store the Wi-Fi firmware patch in external flash.
```
This entry is still in a wrong place (under Matter template sample). I think it needs to be after the Added the Matter manufacturer-specific sample bullet, starting like I suggest above..

ArekBalysNordic replied:
ahhh, ok right

peknis commented on samples/matter/template/README.rst:
```diff
@@ -168,6 +168,21 @@ Building and running
+
+.. important::
+
+   This feature does not work with tools from the `nRF Command Line Tools`_ product page and nrfjprog programming tool.
```
```suggestion
   This feature does not work with tools from the `nRF Command Line Tools`_ product page and the nrfjprog programming tool.
```

peknis reviewed (approved):
Approved with a couple of nits and one more suggestion that might improve the readability.

ArekBalysNordic commented:
@nrfconnect/ncs-co-build-system @nrfconnect/ncs-co-networking @nrfconnect/ncs-wifi-doc Could you please review?

ArekBalysNordic commented on applications/matter_bridge/Kconfig.sysbuild:
```diff
@@ -41,7 +41,7 @@ config PM_EXTERNAL_FLASH_MCUBOOT_SECONDARY
 if SOC_SERIES_NRF53X
 
 config MCUBOOT_UPDATEABLE_IMAGES
-	default 3 if (BOARD_NRF5340DK_NRF5340_CPUAPP && WIFI_NRF70) #### Support for WiFi patch on external flash memory.
+	default 3 if WIFI_NRF70
 	default 2
```
Oh, so something changed then because it didn't work some time ago. Thanks for spotting it. I will remove it from everywhere.

richabp commented on doc/nrf/app_dev/device_guides/nrf70/nrf70_fw_patch_update.rst:
```diff
@@ -161,47 +167,79 @@ Configuring build system
+            .. group-tab:: nRF Connect for VS Code
+
+                    1. When `building an application <How to build an application_>`_ as described in the |nRFVSC| documentation, follow the steps for setting up the build configuration.
+                    #. In the Add Build Configuration screen, click the Add argument button under the Extra CMake argument section.
```
Is it Extra CMake argument**s**?
```suggestion
                    #. In the **Add Build Configuration** screen, click the **Add argument** button under the **Extra CMake argument** section.
```

richabp commented on doc/nrf/app_dev/device_guides/nrf70/nrf70_fw_patch_update.rst:
```diff
@@ -161,47 +167,79 @@ Configuring build system
-        #. In the Add Build Configuration screen, click the Add argument button under the Extra CMake argument section.
-        #. Add the following Kconfig options:
+                    1. When `building an application <How to build an application_>`_ as described in the |nRFVSC| documentation, follow the steps for setting up the build configuration.
+                    #. In the Add Build Configuration screen, click the Add argument button under the Extra CMake argument section.
```
same as above.

richabp commented on doc/nrf/app_dev/device_guides/nrf70/nrf70_fw_patch_update.rst:
```diff
@@ -161,47 +167,79 @@ Configuring build system
 
-* For the nRF5340 DK without the network core: ``2``
-* For the nRF5340 DK with the network core: ``3``
+* For the nRF5340 DK, and nRF7002 DK without the network core: ``2``
+* For the nRF5340 DK, and nRF7002 DK with the network core: ``3``
```
```suggestion
* For the nRF5340 DK and nRF7002 DK without the network core: ``2``
* For the nRF5340 DK and nRF7002 DK with the network core: ``3``
```

richabp commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -642,6 +646,14 @@ Matter samples
 --------------
 
 * Added :ref:`matter_manufacturer_specific_sample` sample that demonstrates an implementation of custom manufacturer-specific clusters used by the application layer.
+* Updated all Matter samples that run Wi-Fi stack to store the Wi-Fi firmware patch in external flash.
+  The change affects the following samples on nRF7002 DK:
```
```suggestion
* Updated all Matter samples that run Wi-Fi stack to store the Wi-Fi firmware patch in external flash.
  The change affects the following samples on the nRF7002 DK:
```

richabp commented on samples/matter/template/README.rst:
```diff
@@ -168,6 +168,27 @@ Building and running
+
+A portion of the application code related to the nRF70 Series' Wi-Fi firmware is saved onto an external memory to free up space in the on-chip memory.
+This option is set by default when building for the nRF7002 DK.
+The additional partition for MCUboot is created in the external memory and it is used to perform a firmware update of the Wi-Fi firmware.
```
```suggestion
An additional partition for MCUboot is created in the external memory and it is used to perform a firmware update of the Wi-Fi firmware.
```

ArekBalysNordic commented:
@nordicjm Thanks for your input. Indeed, setting MCUBOOT_UPDATEABLE_IMAGES is redundant. I've removed it from all places and also fixed the documentation. Could you please look at the current changes?

nordicjm commented on doc/nrf/app_dev/device_guides/nrf70/nrf70_fw_patch_update.rst:
```diff
@@ -161,47 +167,79 @@ Configuring build system
 
-* For the nRF5340 DK with the network core:
+        1. Set the :kconfig:option:`CONFIG_NRF_WIFI_FW_PATCH_DFU` Kconfig option to ``y``.
+        #. Set the ``SB_CONFIG_MCUBOOT_UPDATEABLE_IMAGES`` Kconfig option to ``2`` if for build without the network core, or to ``3`` if the network core is used.
```
line needs removing

nordicjm commented on doc/nrf/app_dev/device_guides/nrf70/nrf70_fw_patch_update.rst:
```diff
@@ -161,47 +167,79 @@ Configuring build system
+
+                    .. code-block:: console
+
+                        west build -d nrf5340dk/nrf5340/cpuapp -d -- -D -DSB_CONFIG_WIFI_PATCHES_EXT_FLASH_STORE=y -DCONFIG_NRF_WIFI_FW_PATCH_DFU=y -DSB_CONFIG_MCUBOOT_UPDATEABLE_IMAGES=3
```
doesn't look fixed, plus updateable images to be dropped

nordicjm commented on doc/nrf/app_dev/device_guides/nrf70/nrf70_fw_patch_update.rst:
```diff
@@ -161,47 +167,79 @@ Configuring build system
+
+                    .. code-block:: console
+
+                        cmake -GNinja -Bbuild -DBOARD=nrf5340dk/nrf5340/cpuapp -DSHIELD=nrf7002ek -DSB_CONFIG_WIFI_PATCHES_EXT_FLASH_STORE=y -DCONFIG_NRF_WIFI_FW_PATCH_DFU=y -DSB_CONFIG_MCUBOOT_UPDATEABLE_IMAGES=3 -DAPP_DIR=*app_path* *path_to_zephyr*/share/sysbuild
```
-updateable

nordicjm commented on samples/matter/template/sysbuild/mcuboot/boards/nrf7002dk_nrf5340_cpuapp.conf:
```diff
@@ -16,7 +16,6 @@ CONFIG_PM_OVERRIDE_EXTERNAL_DRIVER_CHECK=y
 
 # The following configurations are required to support simultaneous multi image update
 CONFIG_PCD_APP=y
```
pcd, upgrade only, flash simulator, zcbor - not needed in these files

ArekBalysNordic replied:
Removed all entries from nrf7002dk files. It should be OK now.

nordicjm reviewed (commented):
seems to be a mixup of changes which are added in first commit then fixed up in second commit

NordicBuilder commented:
### Memory footprint analysis revealed the following potential issues

***sample.matter.template.release[nrf7002dk/nrf5340/cpuapp]***: ***RAM*** size  ***increased by 6120[B]*** in comparison to the ***main[b62a0ed4]*** branch. - [link](https://ci-health.nordicsemi.no/grafana/d/footprint-tracking-sample/footprint-tracking-for-samples-and-applications-sdk-nrf?orgId=1&var-module=sample&var-submodule=matter&var-scenario=template&var-testcase=release&var-arch=arm&var-platform=nrf7002dk/nrf5340/cpuapp) (cc: @kkasperczyk-no @ArekBalysNordic @markaj-nordic)
***sample.matter.template.debug[nrf52840dk/nrf52840]***: ***RAM*** size  ***increased by 5228[B]*** in comparison to the ***main[b62a0ed4]*** branch. - [link](https://ci-health.nordicsemi.no/grafana/d/footprint-tracking-sample/footprint-tracking-for-samples-and-applications-sdk-nrf?orgId=1&var-module=sample&var-submodule=matter&var-scenario=template&var-testcase=debug&var-arch=arm&var-platform=nrf52840dk/nrf52840) (cc: @kkasperczyk-no @ArekBalysNordic @markaj-nordic)
***sample.matter.template.debug[nrf5340dk/nrf5340/cpuapp]***: ***RAM*** size  ***increased by 5228[B]*** in comparison to the ***main[b62a0ed4]*** branch. - [link](https://ci-health.nordicsemi.no/grafana/d/footprint-tracking-sample/footprint-tracking-for-samples-and-applications-sdk-nrf?orgId=1&var-module=sample&var-submodule=matter&var-scenario=template&var-testcase=debug&var-arch=arm&var-platform=nrf5340dk/nrf5340/cpuapp) (cc: @kkasperczyk-no @ArekBalysNordic @markaj-nordic)
***sample.matter.template.release[nrf52840dk/nrf52840]***: ***RAM*** size  ***increased by 5228[B]*** in comparison to the ***main[b62a0ed4]*** branch. - [link](https://ci-health.nordicsemi.no/grafana/d/footprint-tracking-sample/footprint-tracking-for-samples-and-applications-sdk-nrf?orgId=1&var-module=sample&var-submodule=matter&var-scenario=template&var-testcase=release&var-arch=arm&var-platform=nrf52840dk/nrf52840) (cc: @kkasperczyk-no @ArekBalysNordic @markaj-nordic)
***sample.matter.template.release[nrf5340dk/nrf5340/cpuapp]***: ***RAM*** size  ***increased by 5228[B]*** in comparison to the ***main[b62a0ed4]*** branch. - [link](https://ci-health.nordicsemi.no/grafana/d/footprint-tracking-sample/footprint-tracking-for-samples-and-applications-sdk-nrf?orgId=1&var-module=sample&var-submodule=matter&var-scenario=template&var-testcase=release&var-arch=arm&var-platform=nrf5340dk/nrf5340/cpuapp) (cc: @kkasperczyk-no @ArekBalysNordic @markaj-nordic)
***sample.matter.template.debug[nrf7002dk/nrf5340/cpuapp]***: ***RAM*** size  ***increased by 6120[B]*** in comparison to the ***main[b62a0ed4]*** branch. - [link](https://ci-health.nordicsemi.no/grafana/d/footprint-tracking-sample/footprint-tracking-for-samples-and-applications-sdk-nrf?orgId=1&var-module=sample&var-submodule=matter&var-scenario=template&var-testcase=debug&var-arch=arm&var-platform=nrf7002dk/nrf5340/cpuapp) (cc: @kkasperczyk-no @ArekBalysNordic @markaj-nordic)


*Note: This message is automatically posted and updated by the CI (latest/sdk-nrf/PR-20528/17)*

ArekBalysNordic commented:
@nordicjm Is it OK now after providing recent changes?

nordicjm commented:
> @nordicjm Is it OK now after providing recent changes?

The first commit is wrong, it adds `-D <nothing>` and adds the updateable image number Kconfig part, these are then removed in the second commit, doesn't really make sense

ArekBalysNordic commented:
> > @nordicjm Is it OK now after providing recent changes?
> 
> The first commit is wrong, it adds `-D <nothing>` and adds the updateable image number Kconfig part, these are then removed in the second commit, doesn't really make sense

Fixed.

ArekBalysNordic commented:
Decided internally to postpone enabling this feature in Matter due to a lack of CI alignments. The wi-fi part can be merged. @nordicjm could you please review this single commit?

NordicBuilder commented:
### Memory footprint analysis revealed the following potential issues

***applications.nrf_desktop.zdebug.uart[nrf54h20dk@0.9.0/nrf54h20/cpuapp]***: ***ROM*** size  ***increased by 724[B]*** in comparison to the ***main[50bf3820]*** branch. - [link](https://ci-health.nordicsemi.no/grafana/d/footprint-tracking-sample/footprint-tracking-for-samples-and-applications-sdk-nrf?orgId=1&var-module=applications&var-submodule=nrf_desktop&var-scenario=zdebug&var-testcase=uart&var-arch=arm&var-platform=nrf54h20dk@0.9.0/nrf54h20/cpuapp) (cc: @MarekPieta)
***applications.nrf_desktop.zrelease[nrf54h20dk@0.9.0/nrf54h20/cpuapp]***: ***ROM*** size  ***increased by 840[B]*** in comparison to the ***main[50bf3820]*** branch. - [link](https://ci-health.nordicsemi.no/grafana/d/footprint-tracking-sample/footprint-tracking-for-samples-and-applications-sdk-nrf?orgId=1&var-module=applications&var-submodule=nrf_desktop&var-scenario=zrelease&var-testcase=none&var-arch=arm&var-platform=nrf54h20dk@0.9.0/nrf54h20/cpuapp) (cc: @MarekPieta)


*Note: This message is automatically posted and updated by the CI (latest/sdk-nrf/PR-20528/23)*
